src/tree.rs

```diff
@@ -126,12 +128,156 @@ pub fn minimum_spanning_tree(
+    while !node_queue.is_empty() {
+        let node = node_queue.pop_front().unwrap();
+        let pop = pops[node.index()];
+
+        // Mark as seen; push to queue if only one unseen neighbor
+        let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+            .graph
+            .neighbors(node)
+            .filter(|node| !seen_nodes.contains(&node.index()))
+            .collect();
+
+        if unseen_neighbors.len() == 1 {
+            // This will be false if root
+            let neighbor = unseen_neighbors[0];
+            pops[neighbor.index()] += pop;
+            let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
+            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+
+            if !node_queue.contains(&neighbor) {
+                node_queue.push_back(neighbor);
+            }
+        } else if unseen_neighbors.is_empty() {
+            // root
+            break;
+        } else {
+            // Not at the leaves of the unseen subgraph
+            continue;
+        }
+
+        // Check if balanced
+        if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+            balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
+        }
+
+        seen_nodes.insert(node.index());
+    }
+
+    balanced_nodes
+}
+
+/// Bipartition graph into two contiguous, population-balanced components using
+/// mst. Assumes that the graph is contiguous. See :func:`~bipartition_tree` for
+/// details on how balance is defined.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. See
+///     :func:`~minimum_spanning_tree` for details.
+/// :param pops: The populations assigned to each node in the graph.
+/// :param float pop_target: The population target to reach when partitioning
+///     the graph.
+/// :param float epsilon: The maximum percent deviation from the pop_target
+///     allowed while still being a valid balanced cut edge.
+///
+/// :returns: A list of tuples, with each tuple representing a distinct
+///     balanced edge that can be cut. The tuple contains the root of one of the
+///     two partitioned subtrees and the set of nodes making up that subtree.
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, pop, target_pop, epsilon)")]
+pub fn bipartition_graph_mst(
+    py: Python,
+    graph: &graph::PyGraph,
+    weight_fn: PyObject,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> PyResult<Vec<(usize, Vec<usize>)>> {
+    let mut balanced_nodes: Vec<(usize, Vec<usize>)> = vec![];
+    let mut mst = (*graph).clone();
+
+    while balanced_nodes.is_empty() {
+        mst.graph.clear_edges();
+        _minimum_spanning_tree(py, graph, &mut mst, Some(weight_fn.clone()), 1.0)?;
```

Comment:
I don't get why we need to recalculate `mst` every time in the while loop (or even why different loops we'll give different results) since the values of all arguments stays the same.

Reply:
We recalculate `mst` every time in the while loop because `bipartition_graph_mst` is a sampling algorithm and reusing `mst` when `_bipartition_tree` fails would change the distribution that we're sampling from.

Reply:
To be clear: `weight_fn` is intended to sample and be non-deterministic (i.e. sampling between the range `[0, 1]`).

Reply:
We might need to rename weight_fn to something else because on every other function weight_fn takes an edge and returns a weight. Which is not the case here

Reply:
I'm not sure what you mean by that. This should be the same `weight_fn` that `minimum_spanning_tree` takes (that is, it takes in an edge and return a random weight). E.g. `weight_fn` can be: `lambda _: random.random()` (i.e. uniformly picks between `[0,1]`). However, there are other variants of the ReCom Markov Chain sampling algorithm that require certain edges to be prioritized/deprioritized like so: `lambda x: random.random() if edge_crosses_county(x) else random.random() + 0.5`. This is used to create redistricting plan sampling methods that respect county or city boundaries.